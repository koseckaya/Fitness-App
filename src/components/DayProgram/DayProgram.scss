@import "src/styles/vars";

.day-program {
  box-sizing: border-box;
    padding: 10px;
    display: flex;
    flex-direction: column;
    gap: 10px;
    border-radius: 8px;
    background-color: $color-background-primary;
    border: 1px solid #c8c8c8;
    margin-bottom: 15px;
}
.day {
  &__title {
    display: flex;
    justify-content: space-between;
  }
  &__number {
    font-weight: 700;
    font-size: 18px;
  }
  &__desc {
    font-weight: 500;
    color: $secondary;
  }
}
.program__videos{
   display: grid;
    width: 100%;
    gap: 15px;
    grid-template-columns: repeat(5, minmax(0px, 1fr));
}