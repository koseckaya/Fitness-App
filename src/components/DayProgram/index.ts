export { default as DayProgram } from './DayProgram';
export type { Props as DayProgramProps } from './DayProgram';
