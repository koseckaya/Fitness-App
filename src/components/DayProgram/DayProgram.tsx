
import { FC, useState, useCallback, useContext, useMemo } from 'react';
import { useUserData } from '../../routes/Profile';
import { ExerciseVideoPrev } from '../ExerciseVideoPrev';
import { UserContext } from '../utils/contexts';
import { updateUserDocFromAuth } from '../utils/firebase/firebase';
import './DayProgram.scss';
import { useEffect } from 'react';



export type Props = {
    className?: string;
    day: number;
    videos: any[];
    programId: number;
};

const DayProgram: FC<Props> = ({ day, videos, programId }: Props) => {
    const { currentUser } = useContext(UserContext);
    const userData = useUserData(currentUser);


    const isUserAuthorized = useMemo(() => {
        const isAuthorize = currentUser?.email ? true : false;
        return isAuthorize;
    }, [currentUser]);

    const [completedVideos, setCompletedVideos] = useState<string[]>([])
    const [completedDays, setCompletedDays] = useState<number[]>([])

    useEffect(() => {
        const startedChallenges = userData?.challenges ? Object.keys(userData?.challenges) : []
        let completedDaysFire: number[] = [];
        if (startedChallenges.includes(`${programId}`) && userData?.challenges) {
            completedDaysFire = userData?.challenges['' + programId]; 
        }        

        let initialVideos: string[] = [];
        if (completedDaysFire?.includes(day)) {
            initialVideos = videos.map((video, index) => `${day}-${index}`);
        }
        setCompletedDays(completedDaysFire);
        setCompletedVideos(initialVideos)
    }, [userData])

    
    const onVideoClick = useCallback((dayString: string, e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
        e.stopPropagation();
        e.preventDefault();
        if (completedVideos.includes(dayString)) {
            const filt = completedVideos.filter(d => d !== dayString);
            setCompletedVideos(filt)
        } else {
            const newComplVideos = [...completedVideos, dayString];
            setCompletedVideos(newComplVideos)
        }
    }, [setCompletedVideos, completedVideos])


    const handleDayCheck = useCallback(async (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
        if (!isUserAuthorized) return
        const currentProgram = '' + programId;
        const challenges = userData?.challenges;
        if(challenges && currentUser) {
            const completedDays =  challenges[currentProgram];
            let newCompletedDays = [...completedDays]
            if (!completedDays?.includes(day)) newCompletedDays.push(day)
            
            const newChallenges = {
                ...challenges, 
                [currentProgram]: newCompletedDays,
            }
            updateUserDocFromAuth(currentUser, { challenges: newChallenges }).then(() => {
                setCompletedDays([day]);
            });
        }
    }, [isUserAuthorized, userData, day, currentUser, programId])
    
    
    const handleDayUncheck = useCallback((e: React.MouseEvent<HTMLDivElement>) => { 
        if (!isUserAuthorized) return
        const challenges = userData?.challenges;
        const currentProgram = '' + programId;
        
         if(challenges && currentUser) {
            const completedDays =  challenges[currentProgram];
            let newCompletedDays = completedDays.filter(d => d !== day)
             
            const newChallenges = {
                ...challenges, 
                [currentProgram]: newCompletedDays,
             }
            updateUserDocFromAuth(currentUser, { challenges: newChallenges }).then(() => {
                setCompletedDays([]);
                setCompletedVideos([]);
            });
        }

    }, [isUserAuthorized, userData, day,currentUser, programId])
    
    const isCompletedDayVideos = () => {
        const isComplete = completedVideos.length === videos.length ? true : false;
        return isComplete
    }

    const dayChallengeComplAndAuthorize = isCompletedDayVideos() && isUserAuthorized 

    return (
        <div className='day-program'>
            <div className='day__title'>
                <span className='day__number'>Day { day }</span>
                <div className='day__desc'>
                    <span className='day__count'> {videos.length} Workouts | </span>
                    <span className='day__duration'> {videos.reduce((acc, video) => acc += video.duration ,0)} Mins</span>
                </div>
            </div>
        
            <div className='program__videos'>
                {
                    videos.map((video, index) => {
                    let dayIndex = `${day}-${index}`;
                    return <ExerciseVideoPrev title={video.title} onVideoClick={onVideoClick} active={completedVideos.includes(dayIndex)}
                        src={video.src} srcImg={video.srcImg} key={index} day={dayIndex} />
                })}
            </div>
            <div onClick={completedDays?.includes(day) ? handleDayUncheck : dayChallengeComplAndAuthorize ? handleDayCheck: undefined }
                className={`button btn-complete 
                ${isCompletedDayVideos() ? 'active' : ''}
                ${(completedDays?.includes(day)) ? 'completed' : ''}`}>
                { (completedDays?.includes(day)) ? `Day ${day} Complete`: `Mark Day ${day} as Complete` }
                
            </div>
            {isCompletedDayVideos() && !isUserAuthorized  &&
                (<div className='day__message'>Log in or register a free account track your schedule progress</div>)}
            
        </div>
    )
}
    


export default DayProgram;
