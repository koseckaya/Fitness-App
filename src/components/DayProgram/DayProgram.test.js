import React from 'react';
import renderer from 'react-test-renderer';
import { DayProgram } from './DayProgram';


const props = {
    day: 1,
    videos: [],
    programId: 1,
    isCompletedDay: false,
    onDayCheck: () => { },
    onDayUncheck: () => { },
};



describe("DayProgram component", () => {
    it('should render component with default props', () => {

        const component = renderer
            .create(<DayProgram {...props}></DayProgram>)
            .toJSON();
        expect(component).toMatchSnapshot();
    });

    it('should render component without prop day', () => {
        const newProps = { ...props, day: 0 }
        const component = renderer
            .create(<DayProgram {...newProps}></DayProgram>)
            .toJSON();
        expect(component).toMatchSnapshot();
    });
    it('should render component with video prop', () => {
        const video = [{
            title: '5 Min Warm Up',
            src: 'https://www.youtube.com/watch?v=7t1qhpDLSi0&ab_channel=ChloeTing',
            srcImg: 'https://static.chloeting.com/videos/63ae100649cf64962cb594e2/f081bd90-8a98-11ed-97ac-4b900c40c009.jpeg',
            duration: 5
        }]
        const newProps = { ...props, ...video }
        const component = renderer
            .create(<DayProgram {...newProps}></DayProgram>)
            .toJSON();
        expect(component).toMatchSnapshot();
    });
    it('should render component without programId prop', () => {
        const newProps = { ...props, programId: 0 }
        const component = renderer
            .create(<DayProgram {...newProps}></DayProgram>)
            .toJSON();
        expect(component).toMatchSnapshot();
    });
    it('should render component with isCompletedDay prop', () => {
        const newProps = { ...props, isCompletedDay: true }
        const component = renderer
            .create(<DayProgram {...newProps}></DayProgram>)
            .toJSON();
        expect(component).toMatchSnapshot();
    });
})
