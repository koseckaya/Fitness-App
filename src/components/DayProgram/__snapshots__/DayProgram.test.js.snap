// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayProgram component should render component with default props 1`] = `
<div
  className="day-program"
>
  <div
    className="day__title"
  >
    <span
      className="day__number"
    >
      Day 
      1
    </span>
    <div
      className="day__desc"
    >
      <span
        className="day__count"
      >
         
        0
         Workouts | 
      </span>
      <span
        className="day__duration"
      >
         
        0
         Mins
      </span>
    </div>
  </div>
  <div
    className="program__videos"
  />
  <div
    className="button btn-complete 
                active
                "
  >
    Mark Day 1 as Complete
  </div>
  <div
    className="day__message"
  >
    Log in or register a free account track your schedule progress
  </div>
</div>
`;

exports[`DayProgram component should render component with isCompletedDay prop 1`] = `
<div
  className="day-program"
>
  <div
    className="day__title"
  >
    <span
      className="day__number"
    >
      Day 
      1
    </span>
    <div
      className="day__desc"
    >
      <span
        className="day__count"
      >
         
        0
         Workouts | 
      </span>
      <span
        className="day__duration"
      >
         
        0
         Mins
      </span>
    </div>
  </div>
  <div
    className="program__videos"
  />
  <div
    className="button btn-complete 
                active
                completed"
    onClick={[Function]}
  >
    Day 1 Complete
  </div>
  <div
    className="day__message"
  >
    Log in or register a free account track your schedule progress
  </div>
</div>
`;

exports[`DayProgram component should render component with video prop 1`] = `
<div
  className="day-program"
>
  <div
    className="day__title"
  >
    <span
      className="day__number"
    >
      Day 
      1
    </span>
    <div
      className="day__desc"
    >
      <span
        className="day__count"
      >
         
        0
         Workouts | 
      </span>
      <span
        className="day__duration"
      >
         
        0
         Mins
      </span>
    </div>
  </div>
  <div
    className="program__videos"
  />
  <div
    className="button btn-complete 
                active
                "
  >
    Mark Day 1 as Complete
  </div>
  <div
    className="day__message"
  >
    Log in or register a free account track your schedule progress
  </div>
</div>
`;

exports[`DayProgram component should render component without programId prop 1`] = `
<div
  className="day-program"
>
  <div
    className="day__title"
  >
    <span
      className="day__number"
    >
      Day 
      1
    </span>
    <div
      className="day__desc"
    >
      <span
        className="day__count"
      >
         
        0
         Workouts | 
      </span>
      <span
        className="day__duration"
      >
         
        0
         Mins
      </span>
    </div>
  </div>
  <div
    className="program__videos"
  />
  <div
    className="button btn-complete 
                active
                "
  >
    Mark Day 1 as Complete
  </div>
  <div
    className="day__message"
  >
    Log in or register a free account track your schedule progress
  </div>
</div>
`;

exports[`DayProgram component should render component without prop day 1`] = `
<div
  className="day-program"
>
  <div
    className="day__title"
  >
    <span
      className="day__number"
    >
      Day 
      0
    </span>
    <div
      className="day__desc"
    >
      <span
        className="day__count"
      >
         
        0
         Workouts | 
      </span>
      <span
        className="day__duration"
      >
         
        0
         Mins
      </span>
    </div>
  </div>
  <div
    className="program__videos"
  />
  <div
    className="button btn-complete 
                active
                "
  >
    Mark Day 0 as Complete
  </div>
  <div
    className="day__message"
  >
    Log in or register a free account track your schedule progress
  </div>
</div>
`;
